cmd/tailscaled/tailscaled.go

```diff
@@ -96,6 +97,56 @@ var subCommands = map[string]*func([]string) error{
 	"debug":                   &debugModeFunc,
 }
 
+// tryWindowsAppDataMigration attempts to move the Windows state storage
+// directory from its old location to the new location. (Issue 2282)
+//
+// Tailscale 1.14 and before stored state under %LocalAppData%
+// (usually "C:\WINDOWS\system32\config\systemprofile\AppData\Local"
+// when tailscaled.exe is running as a non-user system service).
+// However it is frequently cleared for almost any reason: Windows
+// updates, System Restore, even various System Cleaner utilities.
+//
+// This function does not return an error; on rename failure, code elsewhere
+// in Tailscale will continue to use the old location as a fallback.
+func tryWindowsAppDataMigration() {
+	if runtime.GOOS != "windows" {
+		return
+	}
+
+	oldDir := filepath.Join(os.Getenv("LocalAppData"), "Tailscale")
+	_, err := os.Stat(oldDir)
+	if os.IsNotExist(err) {
+		// Common case.
+		return
+	}
+	if err != nil {
+		log.Printf("tryWindowsAppDataMigration failed; AppDataLocal: %v", err)
+		return
+	}
+
+	programData := filepath.Join(os.Getenv("ProgramData"), "Tailscale")
+	_, err = os.Stat(programData)
+	if err == nil {
+		// Both AppData\Local\Tailscale and ProgramData\Tailscale exist. There are
+		// cases where this can happen, like if one downgrades from a version which
+		// stores state in ProgramData back to an earlier version that re-creates
+		// state in AppData\Local.
+		// This is unusual so we log it, but don't otherwise mess with the files.
+		log.Printf("tryWindowsAppDataMigration ProgramData and AppDataLocal both exist")
+		return
+	} else if !os.IsNotExist(err) {
+		log.Printf("tryWindowsAppDataMigration failed; ProgramData: %v", err)
+		return
+	}
+
+	// Atomically move AppData\Local\Tailscale to ProgramData\Tailscale
+	err = os.Rename(oldDir, programData)
+	if err != nil {
+		log.Printf("tryWindowsAppDataMigration: %v", err)
+		return
+	}
```

Comment:
log.Printf("tryWindowsAppDataMigration: successfully migrated: from %v to %v", oldDir, programData)

Reply:
Done.